// @kedacom3/base-common/src/index.js
import CONSTANT from './constant';
import Base from './base';
import './styles/index.scss';

function jqueryUiShim() {
  $.fn.menu = null; //兼容jquery UI，将jquery UI中的menu禁止掉
  if (!!$.fn.datagrid) {
    $.fn.datagrid.defaults.finder.getTr = function (t, e, a, i) {
      a = a || "body",
        i = i || 0;
      var d = $.data(t, "datagrid"),
        r = d.dc,
        n = d.options;
      if (0 == i) {
        var o = n.finder.getTr(t, e, a, 1),
          l = n.finder.getTr(t, e, a, 2);
        return o.add(l)
      }
      if ("body" == a) {
        var s = $("#" + d.rowIdPrefix + "-" + i + "-" + e);
        return s.length || (s = (1 == i ? r.body1 : r.body2).find("table>tbody>tr[datagrid-row-index=" + e + "]")),
          s
      }
      return "footer" == a ? (1 == i ? r.footer1 : r.footer2).find("table>tbody>tr[datagrid-row-index=" + e + "]") : "selected" == a ? (1 == i ? r.body1 : r.body2).find("table>tbody>tr.datagrid-row-selected") : "highlight" == a ? (1 == i ? r.body1 : r.body2).find("table>tbody>tr.datagrid-row-over") : "checked" == a ? (1 == i ? r.body1 : r.body2).find("table>tbody>tr.datagrid-row-checked") : "editing" == a ? (1 == i ? r.body1 : r.body2).find("table>tbody>tr.datagrid-row-editing") : "last" == a ? (1 == i ? r.body1 : r.body2).find("table>tbody>tr[datagrid-row-index]:last") : "allbody" == a ? (1 == i ? r.body1 : r.body2).find("table>tbody>tr[datagrid-row-index]") : "allfooter" == a ? (1 == i ? r.footer1 : r.footer2).find("table>tbody>tr[datagrid-row-index]") : void 0
    }
  }
}

function initInstant(obj, id, config, data, name) {
  if (typeof config == CONSTANT.ISSTRING) {
    let object = $(id).data(name);
    return fnCallback(object, config, data);
  } else {
    let object = null;
    if (data) {
      object = new obj(id, config, data);
    } else {
      object = new obj(id, config);
    }
    $(id).data(name, object);
    return object;
  }
}

function fnCallback(obj, method, data) {
  if (typeof obj[method] == CONSTANT.ISFUNCTION) {
    try {
      return obj[method](data)
    } catch (e) {
      console.error(e)
    }
  }
  return null;
}
let commFn = {
  isSupportScrollerUI:function(){
    let theUA = window.navigator.userAgent.toLowerCase();
    return theUA.indexOf("webkit") > -1 && theUA.indexOf("edge") < 0;
  }
}

export {
  CONSTANT,
  jqueryUiShim,
  initInstant,
  commFn,
  Base
}
