// @kedacom3/base-scrollBar/webpack.common.js
const path = require('path');

module.exports = {
    entry: {
        ScrollBar: './src/index.js',
    },
    output: {
        path: path.resolve(__dirname, 'dist'),
        filename: 'base-[name].js',
        library: 'BaseScrollBar',
        libraryTarget: 'umd'
    },
    module: {
        rules: [
            {
                test: /\.css$/,
                use: [
                    'style-loader',
                    {
                        loader: 'css-loader',
                        options: {
                            importLoaders: 1
                        }
                    },
                    'postcss-loader'
                ],
            },
            {
                test: /\.js$/,
                include: [
                    path.resolve(__dirname, 'src'),
                    path.resolve(__dirname, 'node_modules/@kedacom')
                ],
                use: {
                    loader: 'babel-loader',
                    options: {
                        cacheDirectory: true,
                        presets: [
                            ['@babel/preset-env',
                                {
                                    'targets': {'ie': '9'}
                                }
                            ]
                        ],
                        plugins: [
                            ['@babel/plugin-transform-runtime',
                                {
                                    'corejs': 2,
                                },
                            ],
                        ],
                    },
                },
            },
            {
                test: /\.scss$/,
                //include:path.resolve(__dirname,'src'),
                use:[
                    'style-loader',
                    {
                        loader:'css-loader',
                        options:{
                            importLoaders:2
                        }

                    },
                    'postcss-loader',
                    {
                        loader: 'sass-loader',
                        options: {
                            implementation: require('sass')
                        }
                    }
                ]
            }
        ],
    },
};
